File: DSA/DSA11001 CÂY BIỂU THỨC 1.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    char val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(char v) : val(v), left(nullptr), right(nullptr) {}
};

void traversal(TreeNode* node)
{
    if (!node)
    {
        return;
    }
    
    traversal(node->left);
    cout << node->val;
    traversal(node->right);
}

void submit()
{
    string s;
    getline(cin, s);

    string temp = "+/*-";
    stack<TreeNode*> A;

    for (int i = 0; i < s.size(); i++)
    {
        if (temp.find(s[i]) == string::npos)
        {
            A.push(new TreeNode(s[i]));
        }
        else
        {
            TreeNode* root = new TreeNode(s[i]);
            root->right = A.top(); A.pop();
            root->left = A.top(); A.pop();
            A.push(root);
        }
    }
    
    TreeNode* root = A.top();

    traversal(root);
    

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}

File: DSA_python/DSA11001 CÂY BIỂU THỨC 1.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right


def traversal(node):
    if node is None:
        return

    traversal(node.left)
    print(node.val, end="")
    traversal(node.right)

def submit():
    s = input()
    temp = "+-/*"
    A = list()

    for i in s:
        if temp.find(i) == -1:
            A.append(TreeNode(i))
        else:
            root = TreeNode(i)
            root.right = A.pop()
            root.left = A.pop()
            A.append(root)

    root = A[-1]

    traversal(root)
    print()





test = int(input())
for i in range(test):
    submit()